src/server/service/kv.rs

```diff
@@ -2636,241 +1279,139 @@ fn future_raw_delete_range<E: Engine, L: LockManager>(
 
 fn future_cop<E: Engine>(
     cop: &Endpoint<E>,
-    req: Request,
     peer: Option<String>,
+    req: Request,
 ) -> impl Future<Item = Response, Error = Error> {
     cop.parse_and_handle_unary_request(req, peer)
         .map_err(|_| unreachable!())
 }
 
-fn extract_region_error<T>(res: &storage::Result<T>) -> Option<RegionError> {
-    use crate::storage::{Error, ErrorInner};
-    match *res {
-        // TODO: use `Error::cause` instead.
-        Err(Error(box ErrorInner::Engine(EngineError(box EngineErrorInner::Request(ref e)))))
-        | Err(Error(box ErrorInner::Txn(TxnError(box TxnErrorInner::Engine(EngineError(
-            box EngineErrorInner::Request(ref e),
-        ))))))
-        | Err(Error(box ErrorInner::Txn(TxnError(box TxnErrorInner::Mvcc(MvccError(
-            box MvccErrorInner::Engine(EngineError(box EngineErrorInner::Request(ref e))),
-        )))))) => Some(e.to_owned()),
-        Err(Error(box ErrorInner::SchedTooBusy)) => {
-            let mut err = RegionError::default();
-            let mut server_is_busy_err = ServerIsBusy::default();
-            server_is_busy_err.set_reason(SCHEDULER_IS_BUSY.to_owned());
-            err.set_server_is_busy(server_is_busy_err);
-            Some(err)
-        }
-        Err(Error(box ErrorInner::GcWorkerTooBusy)) => {
-            let mut err = RegionError::default();
-            let mut server_is_busy_err = ServerIsBusy::default();
-            server_is_busy_err.set_reason(GC_WORKER_IS_BUSY.to_owned());
-            err.set_server_is_busy(server_is_busy_err);
-            Some(err)
-        }
-        Err(Error(box ErrorInner::Closed)) => {
-            // TiKV is closing, return an RegionError to tell the client that this region is unavailable
-            // temporarily, the client should retry the request in other TiKVs.
-            let mut err = RegionError::default();
-            err.set_message("TiKV is Closing".to_string());
-            Some(err)
+macro_rules! make_future {
```

Comment:
IMO, it is clearer to read the macro uses than sift through lots of duplicated boilerplate code - it is hard to pick out the patterns and what is different between them. They're also error prone for the same reason. This module is essentially very simple, it just handles requests and forwards them to the internals, but it is huge and difficult to understand because of the boilerplate.